python/mlir_structured/dialects/indexing.py

```diff
@@ -503,8 +500,32 @@ def _is_scalar(e: Any) -> bool:
   return isinstance(e, Scalar) or isinstance(e, (int, float, bool))
 
 
-def _canonicalize_tuple_index(idx, rank: int):
-  """Helper to remove Ellipsis and replace with implicit trailing slice(None)s.
+def _is_index_tensor(x):
+  """Returns True if x is a Tensor with index dtype, False otherwise."""
+  return (isinstance(x, Value) and Tensor.isinstance(x) and
+          IndexType.isinstance(x.dtype))
+
+
+def _is_int_arraylike(x):
+  """Returns True if x is array-like with integer dtype, False otherwise."""
+  return (isinstance(x, int) and not isinstance(x, bool) or
+          isinstance(x, (list, tuple)) and all(_is_int_arraylike(e) for e in x))
```

Comment:
This allows nested lists. That isn't intended, is it?

Reply:
Why not? You should be able to pass indexers that have shapes like `2x2x2` right? Just btw, in general all of these helper methods should be sensible (I assume) because Jax aims to support numpy advanced indexing as authenticall as possible.

Reply:
Generally, as I look at this PR and its dependent, I would find it helpful to have more doc comments in the form of examples that also hint at the usage, similarly to how we'd write things in MLIR ODS descriptions.

Here we are just talking about supporting underlying mechanics but I already find the polymorphism a bit hard to follow. Just listing the accepted forms in the doc would be useful to me, even if it quickly turns redundant.